Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -927,11 +950,17 @@ public class BottomUpVisitor {
     }
   }
   public class TopDownVisitor<State> {
+    protected bool preResolve = false;
+
     public void Visit(Expression expr, State st) {
       Contract.Requires(expr != null);
       if (VisitOneExpr(expr, ref st)) {
-        // recursively visit all subexpressions and all substatements
-        expr.SubExpressions.Iter(e => Visit(e, st));
+        if (preResolve && expr is ConcreteSyntaxExpression concreteSyntaxExpression) {
+          concreteSyntaxExpression.PreResolveSubExpressions.Iter(e => Visit(e, st));
```

Comment:
>I feel the "two state" nature of our current AST, with its pre and post resolve state, is more complex than it needs to be. I'm hoping that at some point we can remove the ResolvedExpression and ResolvedStatement concepts from our AST.

+ 1

> How would you feel about cloning the AST after parsing, so you can use that instead of having to define and implement PreResolveSubExpressions ?
> Making a clone has a small performance penalty, but I expect us to make post-parsing clones anyways once we start adding caching to our pipeline, so that files which have not been changed but need to be re-resolved because their imports changed, don't have to be parsed again.

I would be all in so that `SubExpressions` is a single field instead of two. The problem is, as [@cpitclaudel  mentioned it](https://github.com/dafny-lang/dafny/pull/2955#discussion_r1010674449), that `.SubExpressions`'s specification is only specified in the comments to list sub expressions _after resolution_, not just after parsing. So there would be an inconsistency.
Maybe for the scope of another PR?